### PR TITLE
Cache ingest shards until head compaction

### LINKED ISSUE #5578: Increasing the number of ingest shards for a tenant can reduce the per-ingester series limit and cause series rejects
#### Describe the bug

When increasing the `ingestion_tenant_shard_size` for a tenant the per-ingester series limit will decrease inversely. This is due to the way we calculate the per-ingester series limit as the `global tenant series limit / ingest shards` since the ingest shards is runtime configurable value that means that a per-ingester series limit that was valid before a shard increase may be no longer valid after ingest shards increase while the tenant remains under the global series limit.
This case can also be triggered by adding ingesters to a cell when shuffle sharding is disabled or a tenant has more ingest shards than the previously available number of ingesters.
Both of these cases will result in series rejected while the tenant is under the global series limit. However the rejects will appear as if the tenant is over the global series limit.

#### To Reproduce
On a Mimir cell with enough available ingest shards/ingesters:
- Have a tenant with a global series limit of 30,000 and 3 ingest shards (10,000 series/ingester limit)
- Write at least 23,000 active series (7,500 series/ingester being written)
- Update the ingest shards to 3 to 6 (5,000 series/ingester new limit)
- Wait until head compaction to clear active series and writes should resume

#### Expected behavior
Since the tenant is under their global series limit Mimir should accept the series.

#### Environment

- Infrastructure: k8s
- Deployment tool: jsonnet

### PR DESCRIPTION
**Pull request by treid314**
#### What this PR does
This PR will cache a tenants `ingestion_tenant_shard_size` until after a head compaction.

#### Which issue(s) this PR fixes or relates to

Fixes #5578

#### Checklist

- [X] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ingester/ingester.go [hunk 1]
```diff
@@ -2443,7 +2443,8 @@ func (i *Ingester) shipBlocks(ctx context.Context, allowed *util.AllowedTenants)
 				level.Error(i.logger).Log("msg", "failed to update cached shipped blocks after shipper synchronisation", "user", userID, "err", err)
 			}
 		}
-
+		// reset shards for tenant table
+		i.limiter.ClearShardsForTenant()
```
Comment (treid314) [inline pkg/ingester/ingester.go:2447]: This introduces a data race, so I'll have to re-think locking around this map in the limiter.

Comment (treid314) [reply pkg/ingester/ingester.go]: This was actually in the wrong place and should not have been in the per user loop

File: pkg/ingester/ingester.go [hunk 2]
```diff
@@ -2383,6 +2391,8 @@ func (i *Ingester) shipBlocks(ctx context.Context, allowed *util.AllowedTenants)
 			return
 		}
 	}
+	// After blocks have been shipped we want to clear the tenant shard count map.
+	defer i.limiter.ClearShardsForTenant()
```
Comment (colega) [inline pkg/ingester/ingester.go:2395]: Why doing this after shipping blocks instead of after compacting them?

**Review by pracucci — COMMENTED** [review]
Thanks Tyler for trying to address this annoying bug! I have few high level comments:
1. Series are dropped from the memory at TSDB head compaction, but a regular TSDB head compaction only drops stale series whose last sample was ingester older than 1h ago (when using the default config). This means that if a series is stale since less than 1h, then you will have to wait 2 regular compactions before it's dropped.
2. I think the proposed solution may work only when you increase the shard size, while it doesn't when shuffle sharding is disabled and you scale out ingesters. I would like to see a solution that works for both cases.

**Comment by treid314** [comment]
This is still a bit of a sketch. Problem 2 should be resolved now though. Problem 1 still requires a little bit more thought.

**Comment by pracucci** [comment]
This PR has been superseeded by #6718 #7087.

**Pull request closed without merging**